File: packages/python/mgx/overlay/webhooks.py
"""Hand-written overlay -- not generated. Inbound webhook verification.

Verify the signature of an inbound MGX webhook before trusting its payload.
The dashboard sender signs with the same scheme implemented here.
"""
from __future__ import annotations

import hashlib
import hmac
import time
from typing import Optional

from mgx.models.webhook_event import WebhookEvent

DEFAULT_TOLERANCE_SECONDS = 300


class MgxSignatureError(Exception):
    """Raised when an inbound webhook signature cannot be verified."""


def _parse_signature_header(header: str) -> tuple[int, str]:
    """Parse ``t=<unix>,v1=<hex hmac>`` into ``(t, v1)``."""
    parts: dict[str, str] = {}
    for segment in header.split(","):
        key, sep, value = segment.partition("=")
        if not sep:
            continue
        parts[key.strip()] = value.strip()
    raw_t = parts.get("t")
    v1 = parts.get("v1")
    if not raw_t or not v1:
        raise MgxSignatureError("Malformed MGX-Signature header.")
    try:
        t = int(raw_t)
    except ValueError as exc:
        raise MgxSignatureError("Malformed MGX-Signature header.") from exc
    return t, v1


def verify_webhook(
    raw_body: str,
    signature_header: str,
    secret: str,
    *,
    tolerance_seconds: int = DEFAULT_TOLERANCE_SECONDS,
    now: Optional[int] = None,
) -> WebhookEvent:
    """Verify an inbound MGX webhook and return the typed event.

    The signature header is ``MGX-Signature: t=<unix>,v1=<hex hmac>``. The HMAC is
    SHA-256 over ``"{t}.{raw_body}"`` keyed with the subscription's signing secret.
    Pass the EXACT raw request body (not a re-serialized object). Raises
    :class:`MgxSignatureError` on a bad signature or a stale timestamp.
    """
    t, v1 = _parse_signature_header(signature_header)
    current = int(time.time()) if now is None else now
    if abs(current - t) > tolerance_seconds:
        raise MgxSignatureError("Webhook timestamp is outside the tolerance window.")
    expected = hmac.new(
        secret.encode(), f"{t}.{raw_body}".encode(), hashlib.sha256
    ).hexdigest()
    if not hmac.compare_digest(expected, v1):
        raise MgxSignatureError("Webhook signature does not match.")
    event = WebhookEvent.from_json(raw_body)
    if event is None:
        raise MgxSignatureError("Webhook body could not be parsed.")
    return event

File: packages/node/overlay/webhooks.ts
/* Hand-written overlay — not generated. Inbound webhook verification. */
import { createHmac, timingSafeEqual } from 'node:crypto'
import type { WebhookEvent } from '../src'

/** Thrown when an inbound webhook signature cannot be verified. */
export class MgxSignatureError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'MgxSignatureError'
  }
}

export interface VerifyOptions {
  /** Max age of the signed timestamp, in seconds. Default 300 (5 minutes). */
  toleranceSeconds?: number
  /** Override "now" (unix seconds) — for testing. */
  now?: number
}

const DEFAULT_TOLERANCE_SECONDS = 300

function parseSignatureHeader(header: string): { t: number; v1: string } {
  const parts: Record<string, string> = {}
  for (const segment of header.split(',')) {
    const idx = segment.indexOf('=')
    if (idx === -1) continue
    parts[segment.slice(0, idx).trim()] = segment.slice(idx + 1).trim()
  }
  const t = Number(parts.t)
  if (!parts.t || Number.isNaN(t) || !parts.v1) {
    throw new MgxSignatureError('Malformed MGX-Signature header.')
  }
  return { t, v1: parts.v1 }
}

/**
 * Verifies an inbound MGX webhook and returns the typed event.
 *
 * The signature header is `MGX-Signature: t=<unix>,v1=<hex hmac>`. The HMAC is
 * SHA-256 over `"{t}.{rawBody}"` keyed with the subscription's signing secret.
 * Pass the EXACT raw request body (not a re-serialized object). Throws
 * {@link MgxSignatureError} on a bad signature or a stale timestamp.
 */
export function verifyWebhook(
  rawBody: string,
  signatureHeader: string,
  secret: string,
  opts: VerifyOptions = {},
): WebhookEvent {
  const { t, v1 } = parseSignatureHeader(signatureHeader)
  const tolerance = opts.toleranceSeconds ?? DEFAULT_TOLERANCE_SECONDS
  const now = opts.now ?? Math.floor(Date.now() / 1000)
  if (Math.abs(now - t) > tolerance) {
    throw new MgxSignatureError('Webhook timestamp is outside the tolerance window.')
  }
  const expected = createHmac('sha256', secret).update(`${t}.${rawBody}`).digest('hex')
  const a = Buffer.from(expected, 'utf8')
  const b = Buffer.from(v1, 'utf8')
  if (a.length !== b.length || !timingSafeEqual(a, b)) {
    throw new MgxSignatureError('Webhook signature does not match.')
  }
  return JSON.parse(rawBody) as WebhookEvent
}
